lib/usdUfe/ufe/UsdUndoDuplicateCommand.cpp

```diff
@@ -0,0 +1,159 @@
+}
+
+} // namespace
+
+UsdUndoDuplicateCommand::UsdUndoDuplicateCommand(
+    const UsdSceneItem::Ptr& srcItem,
+    const UsdSceneItem::Ptr& dstParentItem)
+    : _ufeDstPath(dstParentItem->path())
+    , _ufeSrcPath(srcItem->path())
+    , _dstStage(dstParentItem->prim().GetStage())
+    , _srcStage(srcItem->prim().GetStage())
+{
+    auto srcPrim = srcItem->prim();
+    auto newName = uniqueChildName(dstParentItem->prim(), srcPrim.GetName());
+    _usdDstPath = dstParentItem->prim().GetPath().AppendChild(TfToken(newName));
+}
+
+UsdUndoDuplicateCommand::Ptr UsdUndoDuplicateCommand::create(
+    const UsdSceneItem::Ptr& srcItem,
+    const UsdSceneItem::Ptr& dstParentItem)
+{
+    return std::make_shared<UsdUndoDuplicateCommand>(srcItem, dstParentItem);
+}
+UsdSceneItem::Ptr UsdUndoDuplicateCommand::duplicatedItem() const
+{
+    Ufe::Path ufeSrcPath;
+
+    if (!_ufeDstPath.getSegments().empty()) {
+        ufeSrcPath = appendToPath(_ufeDstPath, _usdDstPath.GetElementString());
+    } else {
+        // Temporary USD stage case.
+        const auto ufePathAsString = "/" + _usdDstPath.GetElementString();
+        ufeSrcPath = Ufe::PathSegment(ufePathAsString, getUsdRunTimeId(), '/');
+    }
+
+    UsdPrim prim = ufePathToPrim(ufeSrcPath, _dstStage);
+    return UsdSceneItem::create(ufeSrcPath, prim);
+}
+
+void UsdUndoDuplicateCommand::execute()
+{
+    UsdUndoBlock undoBlock(&_undoableItem);
+
+    UsdPrim prim = ufePathToPrim(_ufeSrcPath, _srcStage);
+
+    auto path = prim.GetPath();
+    auto stage = prim.GetStage();
+
+    OperationEditRouterContext ctx(EditRoutingTokens->RouteDuplicate, prim);
+
+    // The loaded state of a model is controlled by the load rules of the stage.
+    // When duplicating a node, we want the new node to be in the same loaded
+    // state.
+    duplicateLoadRules(*stage, path, _usdDstPath);
+
+    const auto dstLayer = _dstStage->GetEditTarget().GetLayer();
+
+    // Make sure all necessary parent exists in the target layer, at least as over,
+    // otherwise SdfCopySepc will fail.
+    SdfJustCreatePrimInLayer(dstLayer, _usdDstPath.GetParentPath());
+
+    // Retrieve the local layers around where the prim is defined and order them
+    // from weak to strong. That weak-to-strong order allows us to copy the weakest
+    // opinions first, so that they will get over-written by the stronger opinions.
+    SdfPrimSpecHandleVector authLayerAndPaths = getDefiningPrimStack(prim);
+    std::reverse(authLayerAndPaths.begin(), authLayerAndPaths.end());
+
+    bool isFirst = true;
+
+    for (const SdfPrimSpecHandle& layerAndPath : authLayerAndPaths) {
+        const auto layer = layerAndPath->GetLayer();
+        const auto localPath = layerAndPath->GetPath();
+        const bool result = isFirst
+            ? SdfCopySpec(layer, localPath, dstLayer, _usdDstPath)
+            : UsdUfe::mergePrims(stage, layer, localPath, _dstStage, dstLayer, _usdDstPath);
```

Comment:
Use new mergePrims helper method I added which takes no options (just uses defaults) and verbosity=None.